### ISSUE TITLE
Provide alternate image registry to dockerhub

### ISSUE THREAD

**Issue by onedr0p** [glauth/glauth#298]
Hello 👋🏼 

In order to combat dockerhub's imposed rate limits it would be great to host these images on ghcr (github container registry) or quay.io.

**Comment by Fusion** [comment]
It is currently under investigation. Several considerations here:

1. It looks like Github itself has some limitations. I am particularly concerned about the "Data transfer out" of 1 GB: https://github.com/features/packages#pricing
2. Quay.io appears to be free: https://access.redhat.com/articles/quayio-help#:~:text=Can%20I%20use%20Quay%20for,what%20we%20can%20to%20help! and seems like an overall better home for FOSS project
3. I also looked into gcr but it seems like it would be difficult to keep costs under control

Note that, ideally, we would need to also host the images we rely on in the same location. If a single dependency is still pulled from Docker hub, Quay or Github builds would also fail, being on the receiving end of the rate limiter.

**Comment by onedr0p** [comment]
I'm not sure what Github is really limiting there, I've had no issues using GHCR with millions of pulls on an open source project.

**Comment by Fusion** [comment]
GHCR is part of GitHub Packages.
According to this page (https://docs.github.com/en/billing/managing-billing-for-github-packages/about-billing-for-github-packages):
![image](https://user-images.githubusercontent.com/28699/220255546-7a6d570d-f2d7-4943-a295-e685e6284932.png)

Maybe this does not apply the way I think it does?

**Comment by onedr0p** [comment]
I'm not entirely sure but if you look around you'll see a ton of OSS using GHCR for free and hosting their images without a problem.

**Comment by onedr0p** [comment]
@Fusion not to poke too much but according to Docker you have 30 days to either pay Docker the $300/year cost or migrate the container elsewhere .

**Comment by Fusion** [comment]
Well, things got significantly worse.

This now needs to be two conversations:
- Clearly, you are correct, we are going to prioritize making images available on Github or elsewhere
- In addition, we would originally have kept making images available in docker hub.

If the latter is not possible anymore, then the ramifications are manifold.
Beyond the questions already being asked, such as "what if my organization is deleted and some malicious actor takes over?" I can also envision a scenario where the images are still available, but frozen in time. This will create an interesting drift for any existing pipeline that pulls the `:latest` tag.

**Comment by Heavybullets8** [comment]
docker.io is lowering their unauthenticated pulls yet again to 10 per hour, which causes stagnation issues during cluster bootstraps and what not. 
https://docs.docker.com/docker-hub/usage/

**Comment by onedr0p** [comment]
@Fusion trying to bring this to your attention.

**Comment by Fusion** [comment]
Thanks!

OK it looks like the landscape has somewhat shifted, for the worse:
- container.io is no more
- quay.io appears to require a subscription, even for "free" repos?
- while API limitations exist, github may now definitely be the best choice

And we could use a github action to automatically mirror to docker hub...

**Comment by onedr0p** [comment]
@Fusion quay is free:

https://quay.io/plans/

<img width="821" alt="Image" src="https://github.com/user-attachments/assets/222d60de-101e-4439-8a35-915b132a6448" />

**Comment by Fusion** [comment]
Hopefully this does the trick:

![Image](https://github.com/user-attachments/assets/d2375c57-52af-4ac7-bc87-feaedf1909c6)

**Comment by onedr0p** [comment]
@Fusion are the packages public? I'm not seeing them there or here https://github.com/orgs/glauth/packages?repo_name=glauth

**Comment by Fusion** [comment]
You are right. I cannot believe I am now fighting github to convince it to let everybody else see them.

**Comment by Fusion** [comment]
OK I think I got it, it was some organization-level security. Please check again.

**Comment by onedr0p** [comment]
I see them now ♥️ It might a good idea to setup a GitHub workflow so they're built and pushed there automatically on release but I'll close this issue for now.
